src/NHibernate/Event/EntityListener.cs

```diff
@@ -0,0 +1,211 @@
+                Session = @event.Session,
+                State = @event.State
+            }, typeof (PreUpdateAttribute));
+            return false;
+        }
+
+        private static void InvokeAndApplyDirtyFields(EventConfig config, MethodInfo method)
+        {
+            Dictionary<string, object> initialFieldValues = null;
+            if (config.ShouldUpdateState)
+            {
+                initialFieldValues = GetFieldValues(config.Entity.GetType(), config.Entity);
+            }
+
+            ParameterInfo[] parameters = method.GetParameters();
+            if (parameters.Length == 0)
+            {
+                method.Invoke(config.Entity, new object[] {});
+            }
+            else if (parameters.Length == 1 && parameters[0].ParameterType == typeof (ISession))
+            {
+                method.Invoke(config.Entity, new object[] {config.Session});
+            }
+            else
+            {
+                throw new ArgumentException(
+                    "Error while invoking attribute event method, only 0 or 1 params (ISession) is supported!");
+            }
+
+            if (config.ShouldUpdateState)
+            {
+                UpdateDirtyFieldValuesInState(config, initialFieldValues);
+            }
+        }
+
+        private void InvokeMethodsWithAttribute(EventConfig config, System.Type attributeType)
+        {
+            var entityType = config.Entity.GetType();
+            if (!_methodsWithListenerAttributeCache.ContainsKey(entityType))
+            {
+                _methodsWithListenerAttributeCache.Add(entityType,
+                                                       new Dictionary<System.Type, IEnumerable<MethodInfo>>
+                                                       {
+                                                           {
+                                                               attributeType,
+                                                               GetMethodsWithListenerAttribute(attributeType, entityType)
+                                                           }
+                                                       });
+            }
+
+            Dictionary<System.Type, IEnumerable<MethodInfo>> cacheDict = _methodsWithListenerAttributeCache[entityType];
+            if (!cacheDict.ContainsKey(attributeType))
+            {
+                cacheDict.Add(attributeType, GetMethodsWithListenerAttribute(attributeType, entityType));
+            }
+
+            cacheDict[attributeType].ForEach(method => InvokeAndApplyDirtyFields(config, method));
+        }
+
+        private static IEnumerable<MethodInfo> GetMethodsWithListenerAttribute(System.Type attributeType,
+                                                                               System.Type entityType)
+        {
+            return entityType
+                .GetMethods()
+                .Where(method => method.GetCustomAttributes(attributeType, false).Length > 0);

```

Comment:
The `Where` part yields a deferred enumerable, which will be re-executed at each usage, seriously reducing the benefits of caching it into `_methodsWithListenerAttributeCache`. A `ToList()` or `ToArray()` should be added.

It will be enumerated by concurrent threads by the way, but enumerations of these types are known to be thread safe, even if not documented as such.